parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -719,32 +720,60 @@ modifyAtM path f b = case Path.uncons path of
     -- step the branch by updating its children according to fixup
     pure $ step (setChildBranch seg child') b
 
-stepAt0 :: Applicative m => Path
-                         -> (Branch0 m -> Branch0 m)
-                         -> Branch0 m -> Branch0 m
-stepAt0 p f = runIdentity . stepAt0M p (pure . f)
-
 -- stepManyAt0 consolidates several changes into a single step
-stepManyAt0 :: (Applicative m, Foldable f)
+stepManyAt0 :: forall f m . (Applicative m, Foldable f)
            => f (Path, Branch0 m -> Branch0 m)
            -> Branch0 m -> Branch0 m
-stepManyAt0 actions b = foldl' (\b (p, f) -> stepAt0 p f b) b actions
-
-stepManyAt0M :: (Monad m, Monad n, Foldable f)
+stepManyAt0 actions b = go (first Path.toList <$> toList actions) b where
+  go :: [([NameSegment], Branch0 m -> Branch0 m)] -> Branch0 m -> Branch0 m
+  go actions b = let 
+    -- combines the functions that apply to this level of the tree
+    currentAction b = foldl' (\b f -> f b) b [ f | ([], f) <- actions ]
+
+    -- groups the actions based on the child they apply to
+    childActions :: Map NameSegment [([NameSegment], Branch0 m -> Branch0 m)]
```

Comment:
What's the `[NameSegment]`?

Reply:
It’s just the Path - notice the initial call to go which converts the paths to lists of NameSegment.

Reply:
Oh, I see.  What do you think about leaving it as `Path` for self-documentation purposes?  Maybe the cons (`<|`) and snoc (`|>`) patterns will help?